custom/auth/ante/burntax.go

```diff
@@ -55,6 +105,50 @@ func (btfd BurnTaxFeeDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate
 		// Compute taxes again.
 		taxes := FilterMsgAndComputeTax(ctx, btfd.TreasuryKeeper, msgs...)
 
+		if currHeight >= WhitelistHeight {
+			for _, msg := range msgs {
+				var whitelistedRecipients []string
+				var whitelistedSigners []string
+				recipientWhitelistCount := 0
+
+				switch v := msg.(type) {
+				case *banktypes.MsgSend:
+					whitelistedRecipients = append(whitelistedRecipients, v.ToAddress)
+				case *banktypes.MsgMultiSend:
+					for _, output := range v.Outputs {
+						whitelistedRecipients = append(whitelistedRecipients, output.Address)
+					}
+				default:
+					// TODO: We might want to return an error if we cannot match the msg types, but as such I think that means we also need to cover MsgSetSendEnabled & MsgUpdateParams
+					// return ctx, sdkerrors.Wrap(sdkerrors.ErrInvalidType, "Unsupported message type")
+				}
+
+				signers := msg.GetSigners()
+
+				for _, signer := range signers {
+					signerAddress := signer.String()
+
+					for _, whitelistEntry := range BurnTaxAddressWhitelist {
+						if strings.EqualFold(signerAddress, whitelistEntry) {
+							whitelistedSigners = append(whitelistedSigners, signerAddress)
```

Comment:
Would it be of benefit performance-wise to add a break of the inner for-loop here? No need to loop through the rest of the whitelist after a match.

Second question: Would it – also performance-wise – make more sense to invert the logic like setting a "not-whitelisted" flag on the first entry of either signers or recipients that is NOT in the whitelist? That would save a lot of checks and loops as least of the tx will be whitelisted.

Reply:
As said I am not a go developer, but I imagined something like this (probably wrong syntax):

```go
				var whitelistedRecipients []string
				var whitelistedSigners []string
				recipientWhitelistCount := 0

				switch v := msg.(type) {
				case *banktypes.MsgSend:
					whitelistedRecipients = append(whitelistedRecipients, v.ToAddress)
				case *banktypes.MsgMultiSend:
					for _, output := range v.Outputs {
						whitelistedRecipients = append(whitelistedRecipients, output.Address)
					}
				default:
					// TODO: We might want to return an error if we cannot match the msg types, but as such I think that means we also need to cover MsgSetSendEnabled & MsgUpdateParams
					// return ctx, sdkerrors.Wrap(sdkerrors.ErrInvalidType, "Unsupported message type")
				}

				signers := msg.GetSigners()

				isWhitelisted := true

				for _, signer := range signers {
					signerAddress := signer.String()

					for _, whitelistEntry := range BurnTaxAddressWhitelist {
						if ! strings.EqualFold(signerAddress, whitelistEntry) {
							isWhitelisted := false
							break 2
						}
					}
				}

				if isWhitelisted == true {
					for _, recipient := range whitelistedRecipients {
						for _, whitelistEntry := range BurnTaxAddressWhitelist {
							if ! strings.EqualFold(recipient, whitelistEntry) {
								isWhitelisted := false
								break 2
							}
						}
					}
				}

				if isWhitelisted == true {
					return next(ctx, tx, simulate)
				}
```

Reply:
Forgot to push my last version a few hours ago. Sorry :/

Reply:
What about the reversed logic?

Reply:
From what I see the current check needs at least 43 loop iterations for each sender and for each recipient which is 86 at least. On most transactions it would be sufficient to stop checking after a single check.

Reply:
I am not a big fan of "jump statements". I prefere just breaking from the inner loop, into the outer loop if a match is found for the current signer/recipient check. :)

Reply:
I have inverted the whitelist check. As such your right, however I doubt the iterations are a big concern, we are probably talking milliseconds :)